### PR TITLE
fix and alternative

### PR DESCRIPTION
**Pull request by aloctavodia**
This PR introduces two changes

* Use posterior predictive samples instead of the posterior of mean function
* Compute the KL-divergence over samples and then average

Nevertheless, I want to use this PR to discuss the implications of the current approach. Assuming I am understanding it correctly, and possible alternatives.

With the changes in this PR, the result I get is this.

index

Which is not the expected result but makes sense given the current code (which is not taking into account that for a regression X-y comes in pairs). Thus, one alternative could be to compute instead the squared differences between the sorted posterior predictive distribution (this is similar to one Wasserstein distance) and is cheap given that the distributions are (kind of) already sorted.

At least for this example this provides solution that looks ok
index

### CODE REVIEW

File: src/kulprit/families.py [hunk 1]
```diff
@@ -48,14 +48,15 @@ def kl_div(self, y_ast, y_perp):
...
+        std_ast, std_perp = torch.std(y_ast, 1), torch.std(y_perp, 1)
         # compute KL divergence using full formula
-        div = (
+        div = torch.mean(
             torch.log(std_perp / std_ast)
             + (std_ast**2 + (mu_ast - mu_perp) ** 2) / (2 * std_perp**2)
-            - 1 / 2
+            - 0.5
         )
+        #div = torch.mean((y_ast - y_perp)**2)
```
Comment (yannmclatchie) [inline src/kulprit/families.py:59]: This Wasserstein metric looks to be similar to what is [currently implemented in `projpred`](https://github.com/stan-dev/projpred/blob/develop/R/extend_family.R#L150), so could be a good starting point especially given the exam shown

**Comment by yannmclatchie** [comment]
As you have pointed out, the draws are already in a state such that sorting them is not necessarily needed, although perhaps later we can revisit this possibility if projections do not match our expectations.

I am in favour of using Kullback-Leibler divergence surrogates for the time being given their low cost and the fact they seem to preserve the structure of the distributions well. The `projpred` package have implemented such surrogates in R [here](https://github.com/stan-dev/projpred/blob/develop/R/extend_family.R), so we can always see what they are using as a benchmark.

The `projpred` package also take into account some weighting of data points when computing these metrics, so later it could be worth understanding if such a modification would provide significantly better projections in this case, and how we might go about implementing it.

**Review by yannmclatchie — APPROVED** [review]
